movebucket_test.go

```diff
@@ -0,0 +1,508 @@
+				return
+			}
+
+			// assert: check subBucket has been deleted from srcBucket, and exists in dstBucket
+			if err := db.View(func(tx *bolt.Tx) error {
+				if !tc.srcBucketIsRootBucket {
+					srcBucket := tx.Bucket([]byte(tc.srcBucketName))
+					if srcBucket == nil {
+						t.Fatalf("src bucket %s does not exist: %v", tc.srcBucketName, berrors.ErrBucketNotFound)
+					}
+
+					srcCur := srcBucket.Cursor()
+					k, _ := srcCur.Seek([]byte(tc.subBucketName))
+					if bytes.Equal([]byte(tc.subBucketName), k) {
+						t.Fatalf("key %q still exists in the srcBucket %s : %v", tc.subBucketName, tc.srcBucketName, berrors.ErrIncompatibleValue)
+					}
+				} else {
+					subBucket := tx.Bucket([]byte(tc.subBucketName))
+					if subBucket != nil {
+						t.Fatalf("expected subBucket '%s' to be moved, but still exist in rootBucket", tc.subBucketName)
+					}
+				}
+
+				if !tc.dstBucketIsRootBucket {
+					dstBucket := tx.Bucket([]byte(tc.dstBucketName))
+					if dstBucket == nil {
+						t.Fatalf("dst bucket %s does not exist: %v", tc.dstBucketName, berrors.ErrBucketNotFound)
+					}
+
+					v := dstBucket.Get([]byte(tc.subBucketName))
+					if v != nil {
+						t.Fatalf("expected nil value, as the key is nested bucket, got %v instead", v)
+					}
+					subBucket := dstBucket.Bucket([]byte(tc.subBucketName))
+					v = subBucket.Get([]byte(tc.subBucketKey))
+					if v == nil {
+						t.Fatalf("expected value %v, but got %v instead", tc.subBucketValue, v)
+					}
+				} else {
+					subBucket := tx.Bucket([]byte(tc.subBucketName))
+					if subBucket == nil {
+						t.Fatalf("dst bucket %s does not exist in rootBucket: %v", tc.subBucketName, berrors.ErrBucketNotFound)
+					}
+
+					v := subBucket.Get([]byte(tc.subBucketKey))
+					if v == nil {
+						t.Fatalf("expected value %v, but got %v instead", tc.subBucketValue, v)
+					}
+				}
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+		})
+	}
+}
+
+func TestBucket_MoveBucket(t *testing.T) {
+	testCases := []struct {
+		name          string
+		srcBucketPath []string
+		dstBucketPath []string
+		bucketToMove  string
+		isInlined     bool
+		expActErr     error
+		expAssertErr  error
+	}{
+		{
+			name:          "happy path",
+			srcBucketPath: []string{"x", "y", "z"},
+			dstBucketPath: []string{"a", "b"},
+			bucketToMove:  "z",
+			isInlined:     false,
+			expActErr:     nil,
+			expAssertErr:  nil,
+		},
+		{
+			name:          "bucketToMove not exist in srcBucket",
+			srcBucketPath: []string{"x", "y"},
+			dstBucketPath: []string{"a", "b"},
+			bucketToMove:  "z",
+			isInlined:     false,
+			expActErr:     berrors.ErrBucketNotFound,
+			expAssertErr:  nil,
+		},
+		{
+			name:          "bucketToMove exist in dstBucket",
+			srcBucketPath: []string{"x", "y", "z"},
+			dstBucketPath: []string{"a", "b", "z"},
+			bucketToMove:  "z",
+			expActErr:     berrors.ErrBucketExists,
```

Comment:
@ahrtr this case fails, because the expErr is not returned 

- src bucket contains subBucket z 
- dst bucket contains subBucket z
- yet, the moving takes place, i expected that the key is found and then 

```
// Return an error if there is an existing key in the destination bucket.
	if bytes.Equal(key, k) {
		if (flags & common.BucketLeafFlag) != 0 {
			return errors.ErrBucketExists
		}
		return fmt.Errorf("key %q already exists in the target bucket: %w", key, errors.ErrIncompatibleValue)
	}
```

this is not the case, using the debugger, i could see that the key(i.e. subBucketName `z`) is not found by the cursor !!!!